lib/seattleflu/id3c/cli/command/etl/redcap_det_scan.py

```diff
@@ -0,0 +1,534 @@
+from uuid import uuid4
+from typing import Any, Callable, Dict, List, Mapping, Match, Optional, Union, Tuple
+from datetime import datetime
+from cachetools import TTLCache
+from id3c.db.session import DatabaseSession
+from id3c.cli.command.etl import redcap_det
+from id3c.cli.command.geocode import get_response_from_cache_or_geocoding
+from id3c.cli.command.location import location_lookup
+from seattleflu.id3c.cli.command import age_ceiling
+from .redcap_map import *
+from .fhir import *
+from . import race
+
+
+LOG = logging.getLogger(__name__)
+
+
+REVISION = 1
+
+REDCAP_URL = 'https://redcap.iths.org/'
+INTERNAL_SYSTEM = "https://seattleflu.org"
+
+PROJECT_ID = 20759
+REQUIRED_INSTRUMENTS = [
+    'screening',
+    'shipping_information',
+    'back_end_mail_scans',
+    'illness_questionnaire'
+    'nasal_swab_collection',
+    'post_collection_data_entry_qc'
+]
+
+
+@redcap_det.command_for_project(
+    "scan",
+    redcap_url = REDCAP_URL,
+    project_id = PROJECT_ID,
+    required_instruments = REQUIRED_INSTRUMENTS,
+    raw_coded_values = True,
+    revision = REVISION,
+    help = __doc__)
+
+def redcap_det_scan(*, db: DatabaseSession, cache: TTLCache, det: dict, redcap_record: dict) -> Optional[dict]:
+    location_resource_entries = locations(db, cache, redcap_record)
+    patient_entry, patient_reference = create_patient(redcap_record)
+
+    if not patient_entry:
+        LOG.warning("Skipping enrollment with insufficient information to construct patient")
+        return None
+
+    encounter_entry, encounter_reference = create_encounter(redcap_record, patient_reference, location_resource_entries)
+
+    if not encounter_entry:
+        LOG.warning("Skipping enrollment with insufficient information to construct an encounter")
+        return None
+
+    specimen_entry, specimen_reference = create_specimen(redcap_record, patient_reference)
+
+    if not specimen_entry:
+        LOG.warning("Skipping enrollment with insufficent information to construct a specimen")
+        return None
+
+    questionnaire_entry = create_questionnaire_response(redcap_record, patient_reference, encounter_reference)
+
+    specimen_observation_entry = create_specimen_observation_entry(specimen_reference, patient_reference, encounter_reference)
+
+    resource_entries = [
+        patient_entry,
+        encounter_entry,
+        questionnaire_entry,
+        specimen_entry,
+        *location_resource_entries,
+        specimen_observation_entry
+    ]
+
+    return create_bundle_resource(
+        bundle_id = str(uuid4()),
+        timestamp = datetime.now().astimezone().isoformat(),
+        source = f"{REDCAP_URL}{PROJECT_ID}/{redcap_record['record_id']}",
+        entries = list(filter(None, resource_entries))
+    )
+
+
+def locations(db: DatabaseSession, cache: TTLCache, record: dict) -> list:
+    """ Creates a list of Location resource entries from a REDCap record. """
+    lodging_options = [
+        'shelter',
+        'afl',
```

Comment:
I think an assisted-living facility is a residence, not lodging.

How are shelters categorized in our existing data sources?

Reply:
I based this off of our exisiting REDCap ETLs for swab and send, should I update those as well?

Reply:
Ah, I missed that. Let's leave it this way for now then. Should be fine since we're going to use the primary-location concept in ID3C anyway to fallback from residence → lodging.